<a href="https://colab.research.google.com/github/segnig/ml4fraud-ecommerce-banking/blob/task-2/modelling_credict.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
FILE_PATH = "/content/drive/MyDrive/10 Academy/Data/creditcard.csv.zip"

## **Importing Library**

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

## **Data Loading**

In [6]:
creditcard_df = pd.read_csv(FILE_PATH)
creditcard_df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

## **Modeling**


In [7]:
df = creditcard_df.copy()

In [8]:
df = df.drop('Time', axis=1)

X, y = df.drop('Class', axis=1), df['Class']

In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [10]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

logreg = LogisticRegression()
logreg.fit(X_train, y_train)

y_pred = logreg.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))



Accuracy: 0.9990344440153085

Confusion Matrix:
 [[56853    11]
 [   44    54]]

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.83      0.55      0.66        98

    accuracy                           1.00     56962
   macro avg       0.91      0.78      0.83     56962
weighted avg       1.00      1.00      1.00     56962



/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [11]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier()
rf.fit(X_train, y_train)

y_pred = rf.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.9995786664794073

Confusion Matrix:
 [[56862     2]
 [   22    76]]

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.97      0.78      0.86        98

    accuracy                           1.00     56962
   macro avg       0.99      0.89      0.93     56962
weighted avg       1.00      1.00      1.00     56962



In [12]:
from lightgbm import LGBMClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, roc_auc_score, recall_score, precision_score, f1_score

lgbm = LGBMClassifier()
lgbm.fit(X_train, y_train)

y_pred = lgbm.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))
print("\nROC AUC Score:", roc_auc_score(y_test, y_pred))
print("\nRecall Score:", recall_score(y_test, y_pred))
print("\nPrecision Score:", precision_score(y_test, y_pred))
print("\nF1 Score:", f1_score(y_test, y_pred))




[LightGBM] [Info] Number of positive: 394, number of negative: 227451
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.070704 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 7395
[LightGBM] [Info] Number of data points in the train set: 227845, number of used features: 29
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.001729 -> initscore=-6.358339
[LightGBM] [Info] Start training from score -6.358339
Accuracy: 0.988185105860047

Confusion Matrix:
 [[56238   626]
 [   47    51]]

Classification Report:
               precision    recall  f1-score   support

           0       1.00      0.99      0.99     56864
           1       0.08      0.52      0.13        98

    accuracy                           0.99     56962
   macro avg       0.54      0.75      0.56     56962
weighted avg       1.00      0.99      0.99     56962


ROC AUC Score: 0.7546997203495917

Recall Score: 0.5204081632653061

Preci

In [13]:
from lightgbm import LGBMClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, roc_auc_score, recall_score, precision_score, f1_score

rf = RandomForestClassifier()
rf.fit(X_train, y_train)

y_pred = rf.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))
print("\nROC AUC Score:", roc_auc_score(y_test, y_pred))
print("\nRecall Score:", recall_score(y_test, y_pred))
print("\nPrecision Score:", precision_score(y_test, y_pred))
print("\nF1 Score:", f1_score(y_test, y_pred))




Accuracy: 0.9995962220427653

Confusion Matrix:
 [[56862     2]
 [   21    77]]

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.97      0.79      0.87        98

    accuracy                           1.00     56962
   macro avg       0.99      0.89      0.93     56962
weighted avg       1.00      1.00      1.00     56962


ROC AUC Score: 0.892839557038347

Recall Score: 0.7857142857142857

Precision Score: 0.9746835443037974

F1 Score: 0.8700564971751412


## Credit Card Fraud Detection Report

This report summarizes the steps taken to build and evaluate machine learning models for credit card fraud detection using the provided dataset.

### 1. Data Loading and Preparation

The dataset was loaded from the specified file path: `/content/drive/MyDrive/10 Academy/Data/creditcard.csv.zip`.

- The 'Time' column was dropped as it is not relevant for the modeling task.
- The data was split into features (X) and the target variable (y), where 'Class' represents the target.
- The dataset was further split into training and testing sets using `train_test_split` with a `test_size` of 0.2 and `random_state=42`.

### 2. Model Training and Evaluation

Three different classification models were trained and evaluated: Logistic Regression, Random Forest Classifier, and LightGBM Classifier. The models were evaluated using Accuracy, Confusion Matrix, and Classification Report. Additionally, for LightGBM and the second Random Forest model, ROC AUC Score, Recall Score, Precision Score, and F1 Score were also calculated.

#### 2.1 Logistic Regression

- A Logistic Regression model was trained on the training data.
- The model achieved an accuracy of **0.9990**.
- The confusion matrix shows:
    - True Negatives (correctly predicted non-fraudulent): 56853
    - False Positives (incorrectly predicted fraudulent): 11
    - False Negatives (incorrectly predicted non-fraudulent): 44
    - True Positives (correctly predicted fraudulent): 54
- The classification report indicates good precision and recall for class 0 (non-fraudulent), but lower performance for class 1 (fraudulent) with a recall of 0.55.

#### 2.2 Random Forest Classifier

- A Random Forest Classifier was trained on the training data.
- The model achieved an accuracy of **0.9996**.
- The confusion matrix shows:
    - True Negatives: 56862
    - False Positives: 2
    - False Negatives: 22
    - True Positives: 76
- The classification report shows improved precision and recall for class 1 compared to Logistic Regression, with a recall of 0.78.

#### 2.3 LightGBM Classifier

- A LightGBM Classifier was trained on the training data.
- The model achieved an accuracy of **0.9882**.
- The confusion matrix shows:
    - True Negatives: 56238
    - False Positives: 626
    - False Negatives: 47
    - True Positives: 51
- The classification report indicates a much lower precision for class 1 (0.08) and a recall of 0.52. The ROC AUC Score was 0.7547, Recall Score was 0.5204, Precision Score was 0.0753, and F1 Score was 0.1316.

#### 2.4 Second Random Forest Classifier Evaluation

- The Random Forest Classifier was evaluated again with additional metrics.
- The model achieved an accuracy of **0.9996**.
- The confusion matrix shows:
    - True Negatives: 56862
    - False Positives: 2
    - False Negatives: 21
    - True Positives: 77
- The classification report shows a precision of 0.97 for class 1 and a recall of 0.79. The ROC AUC Score was 0.8928, Recall Score was 0.7857, Precision Score was 0.9747, and F1 Score was 0.8701.

### 3. Summary of Results

Based on the evaluation metrics, the Random Forest Classifier performed the best among the three models, achieving the highest accuracy and significantly better precision, recall, ROC AUC score, and F1 score for detecting fraudulent transactions (Class 1) compared to Logistic Regression and LightGBM. The imbalanced nature of the dataset (very few fraudulent transactions) makes metrics like precision, recall, and F1 score more informative than accuracy alone.

The Random Forest model with the second evaluation metrics appears to be the most promising for this fraud detection task due to its ability to correctly identify a higher percentage of fraudulent transactions (higher recall) while maintaining a high rate of correctly predicted non-fraudulent transactions (high precision for class 0).